Zend/zend_language_parser.y

```diff
@@ -299,10 +300,39 @@ top_statement:
 	|	T_NAMESPACE { RESET_DOC_COMMENT(); }
 		'{' top_statement_list '}'
 			{ $$ = zend_ast_create(ZEND_AST_NAMESPACE, NULL, $4); }
-	|	T_USE use_declarations ';'				{ $$ = $2; $$->attr = T_CLASS; }
-	|	T_USE T_FUNCTION use_declarations ';'	{ $$ = $3; $$->attr = T_FUNCTION; }
-	|	T_USE T_CONST use_declarations ';'		{ $$ = $3; $$->attr = T_CONST; }
-	|	T_CONST const_list ';'					{ $$ = $2; }
+	|	T_USE mixed_batch_use_declarations ';'		{ $$ = $2; }
+	|	T_USE T_FUNCTION batch_use_declarations ';'	{ $$ = $3; $3->attr = T_FUNCTION; }
+	|	T_USE T_CONST batch_use_declarations ';'	{ $$ = $3; $3->attr = T_CONST; }
+	|	T_USE use_declarations ';'					{ $$ = $2; $$->attr = T_CLASS; }
+	|	T_USE T_FUNCTION use_declarations ';'		{ $$ = $3; $$->attr = T_FUNCTION; }
+	|	T_USE T_CONST use_declarations ';'			{ $$ = $3; $$->attr = T_CONST; }
+	|	T_CONST const_list ';'						{ $$ = $2; }
+;
+
+batch_use_declarations:
+	namespace_name T_NS_SEPARATOR '{' use_declarations '}'
+		{$$ = zend_ast_create(ZEND_AST_BATCH_USE, $1, $4); }
+;
+
+mixed_batch_use_declarations:
+	namespace_name T_NS_SEPARATOR '{' inline_use_declarations '}'
+		{$$ = zend_ast_create(ZEND_AST_BATCH_USE, $1, $4);}
+;
+
+inline_use_declarations:
+		inline_use_declarations ',' inline_use_declaration
+			{ $$ = zend_ast_list_add($1, $3); }
+	|	inline_use_declaration
+			{ $$ = zend_ast_create_list(1, ZEND_AST_USE, $1); }
+;
+
+inline_use_declaration:
+		use_declaration
+			{ $$ = $1; $$->attr = T_CLASS; }
+	|	T_FUNCTION use_declaration
+			{ $$ = $2; $$->attr = T_FUNCTION; }
+	|	T_CONST use_declaration
+			{ $$ = $2; $$->attr = T_CONST; }
```

Comment:
I'd suggest adding an `/*empty*/ | T_FUNCTION | T_CONST` rule for determining the attr. It's repeated three times in this patch.